crates/goose/src/agent.rs

```diff
@@ -59,37 +96,136 @@ impl Agent {
+        let message_use = Message::new(
+                Role::Assistant,
+        vec![Content::tool_request_success("000", "status", json!({}))]
+        ).map_err(|e| AgentError::Internal(e.to_string()))?;
+
+        // Create tool result message with status
+        let message_result = Message::new(
+            Role::User,
+            vec![Content::tool_response_success(
+                "000",
+                serde_json::json!(status),
+            )],
+        ).map_err(|e| AgentError::Internal(e.to_string()))?;
+
+        messages.push(message_use);
+        messages.push(message_result);
+        Ok(())
+    }
+
+    // Rewrite the exchange as needed after each tool call
+    // NOTE in this implementation, we do system by system updates other agents might use a very different approach
+    // we from a message list that always looks like:
+    // [kickoff, tool_use_0, tool_result_0, ..., tool_use_n, tool_result_n, status_use, status_result]
+    // where status contains system detail that we always want to include for the agent
+    async fn rewrite_messages_loopend(&self, messages: &mut Vec<Message>, pending: Vec<Message>) -> AgentResult<()> {
+        // Remove the last two messages (status and tool response)
+        messages.pop();
+        messages.pop();
+
+        // Append the pending messages
+        for message in pending {
+            messages.push(message);
+        }
+
+        Ok(())
+    }
+
     /// Create a stream that yields each message as it's generated by the agent.
     /// This includes both the assistant's responses and any tool responses.
     pub fn reply(&self, messages: &[Message]) -> BoxStream<'_, Result<Message>> {
         let mut messages = messages.to_vec();
         let tools = self.get_prefixed_tools();
-        
+        let system_prompt = match self.get_system_prompt() {
+            Ok(prompt) => prompt,
+            Err(e) => return Box::pin(stream::once(async move { Err(e.into()) })),
+        };
+
         Box::pin(async_stream::try_stream! {
             loop {
+                // Append the status tool request and response messages
+                let status = self.get_system_status().await.map_err(|e| AgentError::Internal(e.to_string()))?;
+                self.rewrite_messages_loopstart(&mut messages, status).await?;
+
                 // Get completion from provider
                 let (response, _) = self.provider.complete(
                     &self.model,
-                    "You are an AI agent called Goose. You use tools of connected systems to solve problems.",
+                    &system_prompt,
                     &messages,
                     &tools,
                     None,
                     None,
                 ).await?;
 
-                // Add the assistant's response to the conversation history
-                messages.push(response.clone());
-                
+                // // Add the assistant's response to the conversation history
```

Comment:
maybe leave a comment here that this is handled below, but remove the commented out push